vimprojects/collectedsrc/lftp-4.0.4/doc/lftp.1
.\"
.\" lftp.1 - Sophisticated ftp program
.\"
.\" This file is part of lftp.
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License , or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; see the file COPYING.  If not, write to
.\" the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" $Id: lftp.1,v 1.137 2009/11/19 12:37:28 lav Exp $
.\"
.\"-------
.\" Sp  space down the interparagraph distance
.\"-------
.de Sp
.sp \\n(Ddu
..
.\"-------
.\" Ds  begin a display, indented .5 inches from the surrounding text.
.\"
.\" Note that uses of Ds and De may NOT be nested.
.\"-------
.de Ds
.Sp
.in +0.5i
.nf
..
.\"-------
.\" De  end a display (no trailing vertical spacing)
.\"-------
.de De
.fi
.in
..
.TH lftp 1 "17 Nov 2009"
.SH NAME
lftp \- Sophisticated file transfer program
.SH SYNTAX
.B lftp
[\fB\-d\fR] [\fB\-e \fIcmd\fR] [\fB\-p \fIport\fR]
[\fB\-u \fIuser\fR[\fB,\fIpass\fR]] [\fIsite\fR]
.br
.BI "lftp \-f " script_file
.br
.BI "lftp \-c " commands
.br
.B lftp \-\-version
.br
.B lftp \-\-help

.SH VERSION
This man page documents lftp version 4.0.4.

.SH "DESCRIPTION"
\fBlftp\fR is a file transfer program that allows sophisticated ftp, http
and other connections to other hosts. If \fIsite\fR is specified then lftp
will connect to that site otherwise a connection has to be established with
the open command.
.PP
\fBlftp\fP can handle several file access methods - ftp, ftps, http, https,
hftp, fish, sftp and file (https and ftps are only available when lftp is
compiled with GNU TLS or OpenSSL library). You can specify the method to use in `open
URL' command, e.g. `open http://www.us.kernel.org/pub/linux'. hftp is
ftp-over-http-proxy protocol. It can be used automatically instead of ftp
if ftp:proxy is set to `http://proxy[:port]'. Fish is a protocol working
over an ssh connection to a unix account. SFtp is a protocol implemented
in ssh2 as sftp subsystem.
.PP
Besides FTP-like protocols, lftp has support for BitTorrent protocol
as `torrent' command. Seeding is also supported.

.PP
Every operation in \fBlftp\fP is reliable, that is any not fatal error is
ignored and the operation is repeated. So if downloading breaks, it
will be restarted from the point automatically. Even if ftp server
does not support REST command, \fBlftp\fP will try to retrieve the file from
the very beginning until the file is transferred completely.

\fBlftp\fP has shell-like command syntax allowing you to launch several
commands in parallel in background (&). It is also possible to group
commands within () and execute them in background. All background jobs
are executed in the same single process. You can bring a foreground
job to background with ^Z (c-z) and back with command `wait' (or `fg' which
is alias to `wait'). To list running jobs, use command `jobs'. Some
commands allow redirecting their output (cat, ls, ...) to file or via
pipe to external command. Commands can be executed conditionally based
on termination status of previous command (&&, ||).
.PP
If you exit \fBlftp\fP when some jobs are not finished yet, \fBlftp\fP will move
itself to nohup mode in background. The same happens when you have a
real modem hangup or when you close an xterm.
.PP
\fBlftp\fP has builtin mirror which can download or update a whole directory
tree. There is also reverse mirror (mirror \-R) which uploads or
updates a directory tree on server. Mirror can also synchronize directories
between two remote servers, using FXP if available.
.PP
There is command `at' to launch a job at specified time in current
context, command `queue' to queue commands for sequential execution
for current server, and much more.
.PP
On startup, \fBlftp\fP executes \fI/etc/lftp.conf\fP and then \fI~/.lftprc\fP and
\fI~/.lftp/rc\fP. You can place aliases and `set' commands there. Some
people prefer to see full protocol debug, use `debug' to turn the
debug on. Use `debug 3' to see only greeting messages and error
messages.
.PP
\fBlftp\fP has a number of settable variables. You can use `set \-a' to see
all variables and their values or `set \-d' to see list of defaults.
Variable names can be abbreviated and prefix can be omitted unless the
rest becomes ambiguous.
.PP
If lftp was compiled with OpenSSL (configure \-\-with\-openssl), then it includes software developed
by the OpenSSL Project for use in the OpenSSL Toolkit. (http://www.openssl.org/)

.SS Commands
.PP

.BI ! " shell command"
.PP
Launch shell or shell command.
.PP
.Ds
!ls
.De
.PP
To do a directory listing of the local host.

.B alias
.RI " [" name " [" value ]]
.PP
Define or undefine alias \fIname\fP. If \fIvalue\fP is omitted, the alias is
undefined, else it takes the value \fIvalue\fP. If no argument is given
the current aliases are listed.
.PP
.Ds
alias dir ls \-lF
alias less zmore
.De

.B anon
.PP
Sets the user to anonymous.  This is the default.

.BR at " \fItime\fP [ \-\- \fIcommand\fP ] "
.PP
Wait until the given time and execute given (optional) command. See also \fBat\fR(1).

.B bookmark
.RI " [" subcommand ]
.PP
The bookmark command controls bookmarks.
.Ds
add <name> [<loc>]	add current place or given location
  			to bookmarks and bind to given name
del <name>		remove bookmark with name
edit			start editor on bookmarks file
import <type>		import foreign bookmarks
list			list bookmarks (default)
.De

.B cache
.RI " [" subcommand ]
.PP
The cache command controls local memory cache.
The following subcommands are recognized:
.Ds
stat			print cache status (default)
on|off			turn on/off caching
flush			flush cache
size \fIlim\fP			set memory limit, -1 means unlimited
expire \fINx\fP		set cache expiration time to \fIN\fP seconds (\fIx\fP=s)
			minutes (\fIx\fP=m) hours (\fIx\fP=h) or days (\fIx\fP=d)
.De
.PP

.BR cat " \fIfiles\fP"
.PP
cat outputs the remote file(s) to stdout.  (See also \fBmore\fR,
\fBzcat\fR and \fBzmore\fR)

.BR cd " rdir"
.PP
Change current remote directory.  The previous remote directory is
stored as `\-'. You can do `cd \-' to change the directory back.
The previous directory for each site is also stored on disk,
so you can do `open site; cd \-' even after lftp restart.

.BR chmod " \fImode files\fP"
.PP
Change permission mask on remote files. The mode must be an octal number.

.BR close " [" \-a "]"
.PP
Close idle connections.  By default only with the current server, use
\-a to close all idle connections.

.BR cls " [" OPTS "] " \fIfiles...\fP
.PP
`cls' tries to retrieve information about specified files or directories
and outputs the information according to format options. The difference between
`ls' and `cls' is that `ls' requests the server to format file listing, and
`cls' formats it itself, after retrieving all the needed information.
See `help cls' for options.

.BR command " \fIcmd args...\fP
.PP
execute given command ignoring aliases.

.BR debug " [" \-o
.IR file "] "  level "|\fBoff\fP"
.PP
Switch debugging to \fIlevel\fP or turn it off.  Use \-o to redirect
the debug output to a file.

.BR echo " [" \-n "] \fIstring\fR"
.PP
guess what it does.

.BR eval " [" -f " \fIformat\fR ] " \fIargs...\fR
.PP
without -f it just executes given arguments as a command. With -f, arguments
are transformed into a new command. The format can contain plain text and
placeholders $0...$9 and $@, corresponding to the arguments.

.BR exit " [" bg "] [" top "]"
.RB [ kill ]
.RI [ code ]
.PP
exit will exit from lftp or move to background if there are active jobs. If
no job is active, \fIcode\fP is passed to operating system as lftp's
termination status. If \fIcode\fP is omitted, the exit code of last
command is used.
.PP
`exit bg' forces moving to background when cmd:move-background is false.
`exit top' makes top level `shell' (internal lftp command executor) terminate.
`exit kill' kills all numbered jobs before exiting. The options can be combined, e.g.
`at 08:00 \-\- exit top kill &' kills all jobs and makes lftp exit at specified time.

.B fg
.PP
Alias for `wait'.

.B find
.RI " [" directory "] "
.PP
List files in the directory (current directory by default) recursively.
This can help with servers lacking ls \-R support. You can redirect output
of this command.

.BR ftpcopy
.PP
Obsolete. Use one of the following instead:
.Ds
get ftp://... \-o ftp://...
get \-O ftp://... file1 file2...
put ftp://...
mput ftp://.../*
mget \-O ftp://... ftp://.../*
.De
or other combinations to get FXP transfer (directly between two ftp servers).
lftp would fallback to plain copy (via client) if FXP transfer cannot be
initiated or ftp:use-fxp is false.

.BR get " [" \-E ]
.RB [ \-a "] [" \-c "] [" \-O
.RB "\fIbase\fP] \fIrfile\fP [" \-o " \fIlfile\fP] ..."
.PP
Retrieve the remote file \fIrfile\fP and store it as the local file
\fIlfile\fP.  If \-o is omitted, the file is stored to local file named as
base name of \fIrfile\fP. You can get multiple files by specifying multiple
instances of \fIrfile\fP (and \-o \fIlfile\fP). Does not expand wildcards, use
\fBmget\fR for that.
.Ds
\-c		continue, reget
\-E		delete source files after successful transfer
\-a		use ascii mode (binary is the default)
\-O <base>	specifies base directory or URL where files should be placed
.De
.PP
Examples:
.Ds
get README
get README \-o debian.README
get README README.mirrors
get README \-o debian.README README.mirrors \-o debian.mirrors
get README \-o ftp://some.host.org/debian.README
get README \-o ftp://some.host.org/debian-dir/ (end slash is important)
.De

.BR get1
.RI [ OPTS "] " rfile
.PP
Transfer a single file. Options:
.Ds
\-o <lfile>	destination file name (default - basename of rfile)
\-c		continue, reget
\-E		delete source files after successful transfer
\-a		use ascii mode (binary is the default)
\-\-source\-region=<from-to>
		transfer specified region of source file
\-\-target\-position=<pos>
		position in target file to write data at
.De

.BR glob " [" \-d "] [" \-a "]"
.RB "[" \-f "] \fIcommand patterns\fP"
.PP
Glob given patterns containing metacharacters and pass result to given command.
E.g. ``glob echo *''.
.Ds
\-f	plain files (default)
\-d	directories
\-a	all types
.De

.B help
[\fIcmd\fP]
.PP
Print help for \fIcmd\fP or if no \fIcmd\fP was specified print a list of
available commands.

.B jobs
.RB [ \-v ]
.PP
List running jobs. \-v means verbose, several \-v can be specified.

.B kill
all|\fIjob_no\fP
.PP
Delete specified job with \fIjob_no\fP or all jobs.
(For \fIjob_no\fP see \fBjobs\fP)

.B lcd
\fIldir\fP
.PP
Change current local directory \fIldir\fP. The previous local
directory is stored as `\-'. You can do `lcd \-' to change the directory back.

.B lpwd
.PP
Print current working directory on local machine.

.B ls
\fIparams\fP
.PP
List remote files. You can redirect output of this command to file or
via pipe to external command.  By default, ls output is cached, to see
new listing use
.B rels
or
.B "cache flush."

.BR mget " [" \-c "] [" \-d "]"
.RB [ \-a "] [" \-E "]"
.RB [ \-O " \fIbase\fP] \fIfiles\fP"
.PP
Gets selected files with expanded wildcards.
.PP
.Ds
\-c		continue, reget.
\-d		create directories the same as file names and get
		the files into them instead of current directory.
\-E		delete source files after successful transfer
\-a		use ascii mode (binary is the default)
\-O <base>	specifies base directory or URL where files should be placed
.De

.B mirror
.RI [ OPTS "] [" source
.RI "[" target "]]"
.PP
Mirror specified source directory to local target directory. If target
directory ends with a slash, the source base name is appended to target
directory name. Source and/or target can be URLs pointing to directories.
.PP
.Ds
\-c, \-\-continue		continue a mirror job if possible
\-e, \-\-delete		delete files not present at remote site
    \-\-delete\-first		delete old files before transferring new ones
    \-\-depth\-first		descend into subdirectories before transferring files
\-s, \-\-allow\-suid		set suid/sgid bits according to remote site
    \-\-allow\-chown	try to set owner and group on files
    \-\-ascii		use ascii mode transfers (implies \-\-ignore\-size)
    \-\-ignore\-time		ignore time when deciding whether to download
    \-\-ignore\-size		ignore size when deciding whether to download
    \-\-only\-missing	download only missing files
    \-\-only\-existing	download only files already existing at target
\-n, \-\-only\-newer	download only newer files (\-c won't work)
    \-\-no\-empty\-dirs	don't create empty directories (implies \-\-depth\-first)
\-r, \-\-no\-recursion	don't go to subdirectories
    \-\-no\-symlinks	don't create symbolic links
\-p, \-\-no\-perms		don't set file permissions
    \-\-no\-umask		don't apply umask to file modes
\-R, \-\-reverse		reverse mirror (put files)
\-L, \-\-dereference	download symbolic links as files
\-N, \-\-newer\-than=SPEC	download only files newer than specified time
    \-\-on\-change=CMD		execute the command if anything has been changed
    \-\-older\-than=SPEC	download only files older than specified time
    \-\-size\-range=RANGE	download only files with size in specified range
\-P, \-\-parallel[=N]	download N files in parallel
    \-\-use-pget[\-n=N]	use pget to transfer every single file
    \-\-loop		loop until no changes found
\-i \fIRX\fP, \-\-include \fIRX\fP	include matching files
\-x \fIRX\fP, \-\-exclude \fIRX\fP	exclude matching files
\-I \fIGP\fP, \-\-include\-glob \fIGP\fP	include matching files
\-X \fIGP\fP, \-\-exclude\-glob \fIGP\fP	exclude matching files
\-v, \-\-verbose[=level]	verbose operation
    \-\-log=FILE		write lftp commands being executed to FILE
    \-\-script=FILE		write lftp commands to FILE, but don't execute them
    \-\-just-print, \-\-dry-run	same as \-\-script=-
    \-\-use-cache		use cached directory listings
-\-Remove\-source\-files	remove files after transfer (use with caution)
\-a			same as \-\-allow-chown \-\-allow-suid \-\-no-umask
.De
.PP
When using \-R, the first directory is local and the second is remote.
If the second directory is omitted, base name of first directory is used.
If both directories are omitted, current local and remote directories are used.
If target directory ends with a slash (except root directory) then base
name of source directory is appended.
.PP
\fIRX\fP is an extended regular expression, just like in \fBegrep\fR(1).
.PP
\fIGP\fP is a glob pattern, e.g. `*.zip'.
.PP
Include and exclude options can be specified multiple times. It means that
a file or directory would be mirrored if it matches an include and does
not match to excludes after the include, or does not match anything
and the first check is exclude. Directories are matched with a slash appended.
.PP
Note that symbolic links are not created when uploading to remote server,
because ftp protocol cannot do it. To upload files the links refer
to, use `mirror \-RL' command (treat symbolic links as files).
.PP
For option \-\-newer\-than you can either specify a file or time specification
like that used by \fBat\fR(1) command, e.g. `now-7days' or `week ago'. If you
specify a file, then modification time of that file will be used.
.PP
Verbosity level can be selected using \-\-verbose=level option or by several
\-v options, e.g. \-vvv. Levels are:
.Ds
0 - no output (default)
1 - print actions
2 - +print not deleted file names (when \-e is not specified)
3 - +print directory names which are mirrored
.De
.PP
\-\-only\-newer turns off file size comparison and uploads/downloads
only newer files even if size is different. By default older files are transferred and replace newer ones.
.PP
You can mirror between two servers if you specify URLs instead of directories.
FXP is used automatically for transfers between ftp servers, if possible.
.PP
Some ftp servers hide dot-files by default (e.g. \fI.htaccess\fP), and show
them only when LIST command is used with \-a option. In such case try to use
`set ftp:list-options \-a'.

.B mkdir
.RB "[" \-p "] " \fIdir(s)\fP
.PP
Make remote directories. If \-p is used, make all components of paths.

.B module
.IR "module " [ " args " ]
.PP
Load given module using \fBdlopen\fR(3) function. If module name does not contain
a slash, it is searched in directories specified by module:path variable.
Arguments are passed to module_init function. See README.modules for technical
details.

.B more
\fIfiles\fP
.PP
Same as `cat \fIfiles\fP | more'. if \fBPAGER\fP is set, it is used as filter.
(See also \fBcat\fR, \fBzcat\fR and \fBzmore\fR)

.BR mput " [" \-c "] [" \-d "]"
.RB [ \-a "] [" \-E "]"
.RB [ \-O " \fIbase\fP] \fIfiles\fP"
.PP
Upload files with wildcard expansion. By default it uses the base name of
local name as remote one. This can be changed by `\-d' option.
.Ds
\-c		continue, reput
\-d		create directories the same as in file names and put the
		files into them instead of current directory
\-E		delete source files after successful transfer (dangerous)
\-a		use ascii mode (binary is the default)
\-O <base>	specifies base directory or URL where files should be placed
.De

.B mrm
\fIfile(s)\fP
.PP
Same as `glob rm'. Removes specified file(s) with wildcard expansion.

.B mv
\fIfile1\fP \fIfile2\fP
.PP
Rename \fIfile1\fP to \fIfile2\fP.

.B nlist
[\fIargs\fP]
.PP
List remote file names

.B open
.RB [ \-e " \fIcmd\fP]"
.BR "[" \-u " \fIuser\fP[", "\fIpass\fP]]"
.BR "[" \-p " \fIport\fP] \fIhost\fP|\fIurl\fP"
.PP
Select an ftp server.

.B pget
.RI [ OPTS ]
.I rfile
.RI [ "\fB-o\fP lfile" ]

Gets the specified file using several connections. This can speed up
transfer, but loads the net and server heavily impacting other users. Use only if
you really have to transfer the file ASAP.
Options:
.Ds
\-c		continue transfer. Requires \fIlfile.lftp-pget-status\fP file.
\-n \fImaxconn\fP	set maximum number of connections (default is taken from pget:default-n setting)
.De
.PP

.B put
.RB [ \-E ]
.RB [ \-a ]
.RB [ \-c ]
.RB [ "\-O \fIbase\fP" ]
.I lfile
.RB [ "\-o \fIrfile\fP" ]
.PP
Upload \fIlfile\fP with remote name \fIrfile\fP. If \-o omitted, the base name
of \fIlfile\fP is used as remote name. Does not expand wildcards, use \fBmput\fR for that.
.Ds
\-o <rfile>	specifies remote file name (default - basename of lfile)
\-c		continue, reput
		it requires permission to overwrite remote files
\-E		delete source files after successful transfer (dangerous)
\-a		use ascii mode (binary is the default)
\-O <base>	specifies base directory or URL where files should be placed
.De

.B pwd
.RB [ \-p ]
.PP
Print current remote URL. Use `\-p' option to show password in the URL.

.B queue
.RB [ \-n " \fInum\fP ] \fIcmd\fP"
.PP
Add the given command to queue for sequential execution. Each site has its own
queue. `\-n' adds the command before the given item in the queue. Don't try to
queue `cd' or `lcd' commands, it may confuse lftp. Instead
do the cd/lcd before `queue' command, and it will remember the place in which
the command is to be done. It is possible to queue up an already running job
by `queue wait <jobno>', but the job will continue execution even if it is not
the first in queue.
.PP
`queue stop' will stop the queue, it will not execute any new commands,
but already running jobs will continue to run. You can use `queue stop' to
create an empty stopped queue. `queue start' will resume queue execution.
When you exit lftp, it will start all stopped queues automatically.
.PP
`queue' with no arguments will either create a stopped queue or print queue
status.

.B queue
.BR "\-\-delete|-d " "[\fIindex or wildcard expression\fP]"
.PP
Delete one or more items from the queue. If no argument is given, the last
entry in the queue is deleted.

.B queue
.BR "\-\-move|-m " "<\fIindex or wildcard expression\fP> [\fIindex\fP]"
.PP
Move the given items before the given queue index, or to the end if no
destination is given.
.PP
.Ds
\-q	Be quiet.
\-v	Be verbose.
\-Q	Output in a format that can be used to re-queue.
	Useful with \-\-delete.
.De
.PP
.Ds
> get file &
[1] get file
> queue wait 1
> queue get another_file
> cd a_directory
> queue get yet_another_file
.De
.PP
.Ds
queue \-d 3		Delete the third item in the queue.
queue \-m 6 4		Move the sixth item in the queue before the fourth.
queue \-m "get*zip" 1	Move all commands matching "get*zip" to the beginning
			of the queue.  (The order of the items is preserved.)
queue \-d "get*zip"	Delete all commands matching "get*zip".
.De

.B quote
\fIcmd\fP
.PP
For FTP - send the command uninterpreted. Use with caution - it can lead to
unknown remote state and thus will cause reconnect. You cannot
be sure that any change of remote state because of quoted command
is solid - it can be reset by reconnect at any time.
.PP
For HTTP - specific to HTTP action. Syntax: ``quote <command> [<args>]''.
Command may be ``set-cookie'' or ``post''.
.Ds
open http://www.site.net
quote set-cookie "variable=value; othervar=othervalue"
set http:post-content-type application/x-www-form-urlencoded
quote post /cgi-bin/script.cgi "var=value&othervar=othervalue" > local_file
.De
.PP
For FISH - send the command uninterpreted. This can be used to execute
arbitrary commands on server. The command must not take input or print ###
at new line beginning. If it does, the protocol will become out of sync.
.Ds
open fish://server
quote find \-name \\*.zip
.De


vimprojects/collectedsrc/lftp-4.0.4/src/lftp.cc
/*
 * lftp and utils
 *
 * Copyright (c) 1996-2008 by Alexander V. Lukyanov (lav@yars.free.net)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/* $Id: lftp.cc,v 1.84 2009/11/17 08:18:37 lav Exp $ */

#include <config.h>

#include "modconfig.h"

#include "trio.h"
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h> // for mkdir()
#include "xstring.h"
#include <fcntl.h>
#include <locale.h>
#include <ctype.h>

#include "xmalloc.h"
#include "alias.h"
#include "CmdExec.h"
#include "SignalHook.h"
#include "GetPass.h"
#include "history.h"
#include "log.h"
#include "DummyProto.h"
#include "ResMgr.h"
#include "LsCache.h"
#include "IdNameCache.h"
#include "LocalDir.h"
#include "ConnectionSlot.h"
#include "misc.h"
#include "ArgV.h"

#include "configmake.h"

#include "lftp_rl.h"
#include "complete.h"


#define top_exec CmdExec::top


void  hook_signals()
{
   SignalHook::DoCount(SIGHUP);
   SignalHook::Ignore(SIGTTOU);
   ProcWait::Signal(true);
}

ResDecl res_save_cwd_history
   ("cmd:save-cwd-history","yes",ResMgr::BoolValidate,ResMgr::NoClosure);
ResDecl res_save_rl_history
   ("cmd:save-rl-history","yes",ResMgr::BoolValidate,ResMgr::NoClosure);
ResDecl res_stifle_rl_history
   ("cmd:stifle-rl-history","500",ResMgr::UNumberValidate,ResMgr::NoClosure);

class ReadlineFeeder : public CmdFeeder, private ResClient
{
   bool tty:1;
   bool ctty:1;
   bool add_newline:1;
   int eof_count;
   xstring_c cmd_buf;
   xstring_c for_history;

   static bool readline_inited;
   void readline_init()
   {
      if(readline_inited)
	 return;
      readline_inited=true;
      lftp_readline_init();
      lftp_rl_read_history();
      if(for_history)
      {
	 lftp_add_history_nodups(for_history);
	 for_history.set(0);
      }
      Reconfig(0);
   }

public:
   ReadlineFeeder(const ArgV *args)
   {
      tty=isatty(0);
      ctty=(tcgetpgrp(0)!=(pid_t)-1);
      add_newline=false;
      eof_count=0;
      if(args && args->count()>1)
	 for_history.set_allocated(args->CombineQuoted());
   }
   virtual ~ReadlineFeeder()
   {
      if(readline_inited)
      {
	 if(res_save_cwd_history.QueryBool(0))
	    cwd_history.Save();
	 if(res_save_rl_history.QueryBool(0))
	    lftp_rl_write_history();
      }
   }
   bool RealEOF()
   {
      return !tty || eof_count>3;
   }

   const char *NextCmd(class CmdExec *exec,const char *prompt)
   {
      if(add_newline)
      {
	 add_newline=false;
	 return "\n";
      }

      ::completion_shell=exec;
      ::remote_completion=exec->remote_completion;

      if(tty)
      {
	 readline_init();

	 if(ctty) // controlling terminal
	 {
	    if(!in_foreground_pgrp())
	    {
	       // looks like we are in background. Can't read from tty
	       exec->Timeout(500);
	       return "";
	    }
	 }

	 SignalHook::ResetCount(SIGINT);
	 cmd_buf.set_allocated(lftp_readline(prompt));
	 xmalloc_register_block(cmd_buf.get_non_const());
	 if(cmd_buf && *cmd_buf)
	 {
	    if(exec->csh_history)
	    {
	       char *history_value0=0;
	       int expanded = lftp_history_expand (cmd_buf, &history_value0);
	       if (expanded)
	       {
		  if(history_value0)
		     xmalloc_register_block(history_value0);
		  xstring_ca history_value(history_value0);

		  if (expanded < 0);
		     fprintf (stderr, "%s\n", history_value.get());

		  /* If there was an error, return nothing. */
		  if (expanded < 0 || expanded == 2)	/* 2 == print only */
		  {
		     exec->Timeout(0);  // and retry immediately
		     return "";
		  }

		  cmd_buf.set_allocated(history_value.borrow());
	       }
	    }
	    lftp_add_history_nodups(cmd_buf);
	 }
	 else if(cmd_buf==0 && exec->interactive)
	    puts("exit");

	 if(cmd_buf==0)
	    eof_count++;
	 else
	    eof_count=0;
      }
      else // not a tty
      {
	 if(exec->interactive)
	 {
	    while(*prompt)
	    {
	       char ch=*prompt++;
	       if(ch!=1 && ch!=2)
		  putchar(ch);
	    }
	    fflush(stdout);
	 }
	 cmd_buf.set_allocated(readline_from_file(0));
      }

      ::completion_shell=0;

      if(cmd_buf && last_char(cmd_buf)!='\n')
      {
	 exec->Timeout(0);
	 add_newline=true;
      }
      return cmd_buf;
   }
   void clear()
      {
	 if(!tty)
	    return;
	 lftp_rl_clear();
      }
   void Reconfig(const char *) {
      lftp_rl_history_stifle(res_stifle_rl_history.Query(0));
   }
};
bool ReadlineFeeder::readline_inited;

#define args	  (parent->args)
#define exit_code (parent->exit_code)
#define output	  (parent->output)
#define session	  (parent->session)
#define eprintf	  parent->eprintf
CMD(history)
{
   enum { READ, WRITE, CLEAR, LIST } mode = LIST;
   const char *fn = NULL;
   static struct option history_options[]=
   {
      {"read",required_argument,0,'r'},
      {"write",required_argument,0,'w'},
      {"clear",no_argument,0,'c'},
      {"list",required_argument,0,'l'},
      {0,0,0,0}
   };

   exit_code=0;
   int opt;
   while((opt=args->getopt_long("+r:w:cl",history_options,0))!=EOF) {
      switch(opt) {
      case 'r':
	 mode = READ;
	 fn = optarg;
	 break;
      case 'w':
	 mode = WRITE;
	 fn = optarg;
	 break;
      case 'c':
	 mode = CLEAR;
	 break;
      case 'l':
	 mode = LIST;
	 break;
      case '?':
	 eprintf(_("Try `help %s' for more information.\n"),args->a0());
	 return 0;
      }
   }

   int cnt = 16;
   if(const char *arg = args->getcurr()) {
      if(!strcasecmp(arg, "all"))
	 cnt = -1;
      else if(isdigit((unsigned char)arg[0]))
	 cnt = atoi(arg);
      else {
	 eprintf(_("%s: %s - not a number\n"), args->a0(), args->getcurr());
	 exit_code=1;
	 return 0;
      }
   }

   switch(mode) {
   case READ:
      if(int err = lftp_history_read(fn)) {
	 eprintf("%s: %s: %s\n", args->a0(), fn, strerror(err));
	 exit_code=1;
      }
      break;

   case WRITE:
      if(int err = lftp_history_write(fn)) {
	 eprintf("%s: %s: %s\n", args->a0(), fn, strerror(err));
	 exit_code=1;
      }
      break;

   case LIST:
      lftp_history_list(cnt);
      break;
   case CLEAR:
      lftp_history_clear();
      break;
   }

   return 0;
}
#undef args
#undef exit_code
#undef output
#undef session
#undef eprintf

static void sig_term(int sig)
{
   time_t t=time(0);
   printf(_("[%lu] Terminated by signal %d. %s"),(unsigned long)getpid(),sig,ctime(&t));
   if(top_exec) {
      top_exec->KillAll();
      alarm(30);
      while(Job::NumberOfJobs()>0) {
	 SMTask::Schedule();
	 SMTask::Block();
      }
   }
   exit(1);
}

static void move_to_background()
{
   // notify jobs
   Job::lftpMovesToBackground_ToAll();
   // wait they do something, but no more than 1 sec.
   SMTask::RollAll(TimeInterval(1,0));
   // if all jobs terminated, don't really move to bg.
   if(Job::NumberOfJobs()==0)
      return;

   fflush(stdout);
   fflush(stderr);

   pid_t pid=fork();
   switch(pid)
   {
   case(0): // child
   {
      SignalHook::Ignore(SIGINT);
      SignalHook::Ignore(SIGQUIT);
      SignalHook::Ignore(SIGHUP);
      SignalHook::Ignore(SIGTSTP);

      const char *home=get_lftp_home();
      if(home)
      {
	 char *log=(char*)alloca(strlen(home)+1+3+1);
	 sprintf(log,"%s",home);
	 if(access(log,F_OK)==-1)
	    strcat(log,"_log");
	 else
	    strcat(log,"/log");

	 int fd=open(log,O_WRONLY|O_APPEND|O_CREAT,0600);
	 if(fd>=0)
	 {
	    dup2(fd,1); // stdout
	    dup2(fd,2); // stderr
	    if(fd!=1 && fd!=2)
	       close(fd);
	 }
	 Log::global->ShowPID();
	 Log::global->ShowTime();
	 Log::global->ShowContext();
      }
      close(0);	  // close stdin.
      open("/dev/null",O_RDONLY); // reopen it, just in case.

#ifdef HAVE_SETSID
      setsid();	  // start a new session.
#endif

      pid=getpid();
      time_t t=time(0);
      SignalHook::Handle(SIGTERM,sig_term);
      printf(_("[%lu] Started.  %s"),(unsigned long)pid,ctime(&t));
      for(;;)
      {
	 SMTask::Schedule();
	 if(Job::NumberOfJobs()==0)
	    break;
	 SMTask::Block();
      }
      t=time(0);
      printf(_("[%lu] Finished. %s"),(unsigned long)pid,ctime(&t));
      return;
   }
   default: // parent
      printf(_("[%lu] Moving to background to complete transfers...\n"),
	       (unsigned long)pid);
      fflush(stdout);
      _exit(0);
   case(-1):
      perror("fork()");
   }
}

int lftp_slot(int count,int key)
{
   if(!top_exec)
      return 0;
   char slot[2];
   slot[0]=key;
   slot[1]=0;
   top_exec->ChangeSlot(slot);
   lftp_rl_set_prompt(top_exec->MakePrompt());
   return 0;
}

void  source_if_exist(const char *rc)
{
   if(access(rc,R_OK)!=-1)
   {
      top_exec->FeedCmd("source ");
      top_exec->FeedCmd(rc);
      top_exec->FeedCmd("\n");
   }
}

static void tty_clear()
{
   if(top_exec)
      top_exec->pre_stdout();
}

char *program_name;

int   main(int argc,char **argv)
{
   program_name=argv[0];

#ifdef SOCKS4
   SOCKSinit(program_name);
#endif

   setlocale (LC_ALL, "");
   setlocale (LC_NUMERIC, "C");
   bindtextdomain (PACKAGE, LOCALEDIR);
   textdomain (PACKAGE);

   CmdExec::RegisterCommand("history",cmd_history,
	 N_("history -w file|-r file|-c|-l [cnt]"),
	 N_(" -w <file> Write history to file.\n"
	 " -r <file> Read history from file; appends to current history.\n"
	 " -c  Clear the history.\n"
	 " -l  List the history (default).\n"
	 "Optional argument cnt specifies the number of history lines to list,\n"
	 "or \"all\" to list all entries.\n"));

   top_exec=new CmdExec(0,0);
   hook_signals();
   top_exec->SetStatusLine(new StatusLine(1));
   Log::global->SetCB(tty_clear);

   source_if_exist(SYSCONFDIR"/lftp.conf");
   const char *home=getenv("HOME");
   if(home)
      source_if_exist(dir_file(home,".lftprc"));
   home=get_lftp_home();
   if(home)
      source_if_exist(dir_file(home,"rc"));
   top_exec->WaitDone();

   top_exec->SetTopLevel();
   top_exec->Fg();

   Ref<ArgV> args(new ArgV(argc,argv));
   args->setarg(0,"lftp");

   lftp_feeder=new ReadlineFeeder(args);

   top_exec->ExecParsed(args.borrow());
   top_exec->WaitDone();
   int exit_code=top_exec->ExitCode();

   top_exec->AtExit();
   top_exec->WaitDone();

   if(Job::NumberOfJobs()>0)
   {
      top_exec->SetInteractive(false);
      move_to_background();
   }
   top_exec->KillAll();
   top_exec=0;

   Job::Cleanup();
   ConnectionSlot::Cleanup();
   SessionPool::ClearAll();
   FileAccess::ClassCleanup();
   ProcWait::DeleteAll();
   IdNameCacheCleanup();
   SignalHook::Cleanup();
   Log::Cleanup();
   SMTask::Cleanup();

   return exit_code;
}
